mantidproject/mantid
# Fix ADS observer behaviour on Fitting tab in Engineering Diffraction UI

**Description of work.**

The fitting tab creates ~3 workspaces when you load a workspace:
a) the focussed workspace the user has loaded
b) a derived workspace that has background subtracted
c) a log group workspace

If the user renamed or deleted any of these the UI would often crash or log a nasty looking error message. The ADS observer behaviour has been reworked to sort this out. This relies on a new class called WorkspaceContainer in the fitting tab's model that links together the a) and b) workspaces in the list above

As part of this I noticed that the serial and sequential fit weren't taking into account the Subtract BG checkbox which indicates whether a) or b) is active for each loaded workspace. So I've fixed this bug also.

**To test:**

Open up the Engineering Diffraction interface
Go to the Fitting Tab
Load a focused data file - you can load the one in the attached zip file
[ENGINX_299080_307521_bank_2_TOF.zip](https://github.com/mantidproject/mantid/files/7556728/ENGINX_299080_307521_bank_2_TOF.zip)
Check the "Plot" and "Subtract BG" checkbox on the row created for this file in the table
Try renaming or deleting the workspaces in ADS and check the UI does something sensible


Fixes #32837.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
